src/main/java/org/kiwiproject/dropwizard/error/dao/jdk/ConcurrentMapApplicationErrorDao.java

```diff
@@ -0,0 +1,277 @@
+
+        var newId = currentId.incrementAndGet();
+        var errorWithId = newError.withId(newId);
+
+        // Ensure it is unresolved
+        var unresolvedError = errorWithId.isResolved() ?
+                updateWith(errorWithId, errorWithId.getNumTimesOccurred(), false) : errorWithId;
+        errors.put(newId, unresolvedError);
+
+        return newId;
+    }
+
+    @Override
+    public void incrementCount(long id) {
+        var error = errors.get(id);
+        checkState(nonNull(error), "Unable to increment count. No ApplicationError found with id %s", id);
+
+        var updatedError = incrementNumTimesOccurred(error);
+        errors.put(id, updatedError);
+    }
+
+    @Override
+    public long insertOrIncrementCount(ApplicationError error) {
+        var id = error.getId();
+        if (isNull(id)) {
+            return insertError(error);
+        }
+
+        incrementCount(id);
+        return id;
+    }
+
+    @Override
+    public ApplicationError resolve(long id) {
+        var error = errors.get(id);
+
+        checkState(nonNull(error), "Unable to resolve. No ApplicationError found with id %s", id);
+
+        var resolvedError = resolve(error);
+        errors.put(id, resolvedError);
+
+        return resolvedError;
+    }
+
+    @Override
+    public int resolveAllUnresolvedErrors() {
+        var count = errors.values()
+                .stream()
+                .filter(byStatus(ApplicationErrorStatus.UNRESOLVED))
+                .map(applicationError -> resolve(applicationError.getId()))
+                .count();
+
+        return Math.toIntExact(count);
+    }
+
+    private static Predicate<ApplicationError> byStatus(ApplicationErrorStatus status) {
+        switch (status) {
+            case ALL:
+                return applicationError -> true;
+
+            case RESOLVED:
+                return ApplicationError::isResolved;
+
+            case UNRESOLVED:
+                return not(ApplicationError::isResolved);
+
+            default:
+                throw new IllegalStateException("unknown status: " + status);
+        }
+    }
+
+    @Override
+    public int deleteResolvedErrorsBefore(ZonedDateTime expirationDate) {
+        return deleteErrorsWithStatusAndBefore(ApplicationErrorStatus.RESOLVED, expirationDate);
+    }
+
+    @Override
+    public int deleteUnresolvedErrorsBefore(ZonedDateTime expirationDate) {
+        return deleteErrorsWithStatusAndBefore(ApplicationErrorStatus.UNRESOLVED, expirationDate);
+    }
+
+    private int deleteErrorsWithStatusAndBefore(ApplicationErrorStatus status, ZonedDateTime referenceDate) {
+        checkArgument(isResolvedOrUnresolved(status));
+
+        // TODO Is there a better way to count than this? (removeIf returns a boolean, which isn't useful here)
+        var removeCount = new AtomicInteger();
+        errors.entrySet().removeIf(entry -> matchesStatusAndBeforeDate(entry, status, referenceDate, removeCount));
```

Comment:
`.stream().filter(b -> b).count()` ?

Reply:
This doesn't actually remove elements, though. The problem is that I need to remove elements, _and_ count the number that were removed. The removeIf method removes elements based on a Predicate, and returns a boolean which is true if any elements were removed. So, it doesn't work here since it doesn't tell me how many elements were removed.

Reply:
I hoped that it returned a stream of boolean values for each entry, oh well.

Reply:
@chrisrohr you have any ideas?

Reply:
No I do not

Reply:
OK, then I will resolve this conversation, but I'll keep the TODO in there, in case I have any brilliant (?) ideas about a cleaner way.